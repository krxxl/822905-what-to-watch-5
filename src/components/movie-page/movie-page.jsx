import React from 'react';
import PropTypes from 'prop-types';
import Header from '../header/header';
import Footer from '../footer/footer';
import MovieList from '../movie-list/movie-list';
import {Link} from "react-router-dom";
import MovieDesc from '../movie-desc/movie-desc';
import tabNames from '../../movie-tabs-names';
import COUNTFILM from '../../constant/constant';
import withActiveItem from '../../hocs/with-active-item/with-active-item';

const Components = withActiveItem(MovieDesc);

const MoviePage = (props) => {

  const {films, reviews, onSmallCardClick, filmId} = props;
  const film = films.find((item)=>item.id === +filmId);
  const filmReviews = reviews.find((item)=>item.id === +filmId);
  const {hero, name, poster, genre, year} = film;

  return (
    <React.Fragment>
      <section className="movie-card movie-card--full">
        <div className="movie-card__hero">
          <div className="movie-card__bg">
            <img src={hero} alt={name} />
          </div>

          <h1 className="visually-hidden">WTW</h1>

          <Header />

          <div className="movie-card__wrap">
            <div className="movie-card__desc">
              <h2 className="movie-card__title">{name}</h2>
              <p className="movie-card__meta">
                <span className="movie-card__genre">{genre}</span>
                <span className="movie-card__year">{year}</span>
              </p>

              <div className="movie-card__buttons">
                <button className="btn btn--play movie-card__button" type="button">
                  <svg viewBox="0 0 19 19" width="19" height="19">
                    <use href="#play-s"></use>
                  </svg>
                  <span>Play</span>
                </button>
                <button className="btn btn--list movie-card__button" type="button">
                  <svg viewBox="0 0 19 20" width="19" height="20">
                    <use href="#add"></use>
                  </svg>
                  <span>My list</span>
                </button>
                <Link to={`/films/${filmId}/review`} className="btn movie-card__button">Add review</Link>
              </div>
            </div>
          </div>
        </div>

        <div className="movie-card__wrap movie-card__translate-top">
          <div className="movie-card__info">
            <div className="movie-card__poster movie-card__poster--big">
              <img src={poster} alt={name} width="218" height="327" />
            </div>
            <Components film={film} filmReviews={filmReviews} tabNames={tabNames}/>
          </div>
        </div>
      </section>

      <div className="page-content">
        <section className="catalog catalog--like-this">
          <h2 className="catalog__title">More like this</h2>
          <MovieList COUNTFILM={COUNTFILM} genre={genre} films={films} onSmallCardClick={onSmallCardClick} />
        </section>

        <Footer />
      </div>
    </React.Fragment>
  );
};

MoviePage.propTypes = {
  filmId: PropTypes.number.isRequired,
  films: PropTypes.arrayOf(PropTypes.shape({
    name: PropTypes.string.isRequired,
    genre: PropTypes.string.isRequired,
    year: PropTypes.string.isRequired,
    poster: PropTypes.string.isRequired,
    preview: PropTypes.string.isRequired,
    hero: PropTypes.string.isRequired,
    time: PropTypes.string.isRequired,
    rating: PropTypes.string.isRequired,
    count: PropTypes.number.isRequired,
    desc: PropTypes.string.isRequired,
    director: PropTypes.string.isRequired,
    starring: PropTypes.array.isRequired,
    isInList: PropTypes.bool.isRequired,
    video: PropTypes.string.isRequired,
    id: PropTypes.number.isRequired,
  }).isRequired).isRequired,
  onSmallCardClick: PropTypes.func.isRequired,
  reviews: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    reviews: PropTypes.arrayOf(PropTypes.shape({
      text: PropTypes.string.isRequired,
      author: PropTypes.string.isRequired,
      date: PropTypes.string.isRequired,
      rating: PropTypes.number.isRequired,
      id: PropTypes.number.isRequired,
    }).isRequired).isRequired,
  })).isRequired
};

export default MoviePage;
