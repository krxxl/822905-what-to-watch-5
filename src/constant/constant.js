export const COUNT_FILM = 4;
export const SHOW_ON_BUTTON_FILMS = 8;
export const SHOW_ON_STAR_FILMS = 8;
export const TIMEOUT = 1000;
export const DEFAULT_GENRE = `All genres`;
export const REG_EXP_MAIL = /^([\w.%+-]+)@([\w-]+\.)+([\w]{2,})$/i;

export const AuthorizationStatus = {
  AUTH: `AUTH`,
  NO_AUTH: `NO_AUTH`,
};

export const Status = {
  TRUE: 1,
  FALSE: 0,
};


export const ErrorStatus = {
  ERROR: `ERROR`,
  NO_ERROR: `NO_ERROR`,
};

export const getFilmById = (state, props) => {
  return state.DATA.films.find((item) => item.id === props.filmId);
};

export const TabName = {
  DETAIL: `Detail`,
  OVERVIEW: `Overview`,
  REVIEW: `Review`
};

export const Rating = {
  BAD: `Bad`,
  NORMAL: `Normal`,
  GOOD: `Good`,
  VERYGOOD: `Very good`,
  AWESOME: `Awesome`
};
