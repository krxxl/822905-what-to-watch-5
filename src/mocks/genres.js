export const genres = [
  {
    name: `All genres`,
    id: 1
  },
  {
    name: `Comedies`,
    id: 2
  },
  {
    name: `Crime`,
    id: 3
  },
  {
    name: `Documentary`,
    id: 4
  },
  {
    name: `Dramas`,
    id: 5
  },
  {
    name: `Horror`,
    id: 6
  },
  {
    name: `Kids & Family`,
    id: 7
  },
  {
    name: `Romance`,
    id: 8
  },
  {
    name: `Sci-Fi`,
    id: 9
  },
  {
    name: `Thrillers`,
    id: 10
  },
];

export const genresInSingular = {
  ALL: `All genres`,
  COMEDY: `Comedies`,
  CRIME: `Crime`,
  DOCUMENTARY: `Documentary`,
  DRAMA: `Dramas`,
  HORROR: `Horror`,
  KIDS: `Kids & Family`,
  ROMANCE: `Romance`,
  SCI_FI: `Sci-Fi`,
  THRILLER: `Thrillers`,
};
