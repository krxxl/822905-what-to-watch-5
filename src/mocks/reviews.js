export default [
  {
    id: 1,
    reviews: [
      {
        text: `It is certainly a magical and childlike way of storytelling, even if the content is a little more adult.`,
        author: `Matthew Lickona`,
        date: `December 24, 2016`,
        rating: 8.9,
        id: 1
      },
      {
        text: `It is certainly a magical and childlike way of storytelling, even if the content is a little more adult.`,
        author: `Paula Fleri-Soler`,
        date: `December 20, 2016`,
        rating: 8.9,
        id: 2
      },
      {
        text: `The mannered, madcap proceedings are often delightful, occasionally silly, and here and there, gruesome and/or heartbreaking.`,
        author: `Bill Goodykoontz`,
        date: `November 18, 2015`,
        rating: 8.9,
        id: 3
      },
    ]
  },
  {
    id: 2,
    reviews: [
      {
        text: `The mannered, madcap proceedings are often delightful, occasionally silly, and here and there, gruesome and/or heartbreaking.`,
        author: `Matthew Lickona`,
        date: `December 20, 2016`,
        rating: 8.9,
        id: 4
      },
      {
        text: `It is certainly a magical and childlike way of storytelling, even if the content is a little more adult.`,
        author: `Bill Goodykoontz`,
        date: `December 24, 2016`,
        rating: 8.9,
        id: 5
      },
    ]
  },
  {
    id: 3,
    reviews: [
      {
        text: `The mannered, madcap proceedings are often delightful, occasionally silly, and here and there, gruesome and/or heartbreaking.`,
        author: `Matthew Lickona`,
        date: `December 24, 2016`,
        rating: 8.9,
        id: 6
      },
    ]
  },
  {
    id: 4,
    reviews: [
      {
        text: `It is certainly a magical and childlike way of storytelling, even if the content is a little more adult.`,
        author: `Matthew Lickona`,
        date: `November 18, 2015`,
        rating: 8.9,
        id: 7
      },
      {
        text: `It is certainly a magical and childlike way of storytelling, even if the content is a little more adult.`,
        author: `Paula Fleri-Soler`,
        date: `December 24, 2016`,
        rating: 8.9,
        id: 8
      },
      {
        text: `The mannered, madcap proceedings are often delightful, occasionally silly, and here and there, gruesome and/or heartbreaking.`,
        author: `Bill Goodykoontz`,
        date: `December 24, 2016`,
        rating: 8.9,
        id: 9
      },
    ]
  },
  {
    id: 5,
    reviews: [
      {
        text: `The mannered, madcap proceedings are often delightful, occasionally silly, and here and there, gruesome and/or heartbreaking.`,
        author: `Matthew Lickona`,
        date: `December 24, 2016`,
        rating: 8.9,
        id: 10
      },
      {
        text: `It is certainly a magical and childlike way of storytelling, even if the content is a little more adult.`,
        author: `Paula Fleri-Soler`,
        date: `December 20, 2016`,
        rating: 8.9,
        id: 11
      },
    ]
  },
  {
    id: 6,
    reviews: [
      {
        text: `The mannered, madcap proceedings are often delightful, occasionally silly, and here and there, gruesome and/or heartbreaking.`,
        author: `Matthew Lickona`,
        date: `December 24, 2016`,
        rating: 8.9,
        id: 12
      },
    ]
  },
  {
    id: 7,
    reviews: [
      {
        text: `It is certainly a magical and childlike way of storytelling, even if the content is a little more adult.`,
        author: `Matthew Lickona`,
        date: `December 24, 2016`,
        rating: 8.9,
        id: 13
      },
      {
        text: `It is certainly a magical and childlike way of storytelling, even if the content is a little more adult.`,
        author: `Paula Fleri-Soler`,
        date: `December 20, 2016`,
        rating: 8.9,
        id: 14
      },
      {
        text: `The mannered, madcap proceedings are often delightful, occasionally silly, and here and there, gruesome and/or heartbreaking.`,
        author: `Bill Goodykoon`,
        date: `November 18, 2015`,
        rating: 8.9,
        id: 15
      },
    ]
  },
  {
    id: 8,
    reviews: [
      {
        text: `The mannered, madcap proceedings are often delightful, occasionally silly, and here and there, gruesome and/or heartbreaking.`,
        author: `Matthew Lickona`,
        date: `December 24, 2016`,
        rating: 8.9,
        id: 16
      },
    ]
  }
];
