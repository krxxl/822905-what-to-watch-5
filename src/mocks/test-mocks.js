import films from './films';
import reviews from './reviews';
import {AuthorizationStatus} from "../constant/constant";
import {SHOW_ON_STAR_FILMS} from '../constant/constant';

export const TEST_MOCK_STORE = {
  DATA: {
    films,
    reviews,
    favoriteFilms: films,
    isLoading: true,
    isPromoLoading: true,
    sendingReview: false,
    sendingReviewError: false,
    promoFilm: films[0],
  },
  USER: {
    authorizationStatus: AuthorizationStatus.NO_AUTH,
  },
  SHOW: {
    genreActive: `All genres`,
    count: SHOW_ON_STAR_FILMS,
  }
};
