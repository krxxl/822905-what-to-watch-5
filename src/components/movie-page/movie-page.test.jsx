import React from 'react';
import renderer from 'react-test-renderer';
import {MoviePage} from './movie-page';
import films from '../../mocks/films';
import {AuthorizationStatus} from '../../constant/constant';
import {Provider} from 'react-redux';
import configureMockStore from 'redux-mock-store';
import {TEST_MOCK_STORE} from '../../mocks/test-mocks';
import thunk from 'redux-thunk';
import {createAPI} from '../../services/api';
import {MemoryRouter} from 'react-router-dom';

const api = createAPI(() => {});

const middlewares = [thunk.withExtraArgument(api)];
const mockStore = configureMockStore(middlewares);
const store = mockStore(TEST_MOCK_STORE);

describe(`Render MoviePage`, () => {
  it(`Render MoviePage`, () => {
    const tree = renderer
      .create(
          <Provider store={store}>
            <MemoryRouter>
              <MoviePage
                filmId={films[0].id}
                films={films}
                authorizationStatus={AuthorizationStatus.AUTH}
              />
            </MemoryRouter>
          </Provider>,
          {
            createNodeMock: () => {
              return {};
            },
          }
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
