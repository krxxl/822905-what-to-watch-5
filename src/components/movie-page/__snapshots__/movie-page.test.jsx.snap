// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render MoviePage Render MoviePage 1`] = `
Array [
  <section
    className="movie-card movie-card--full"
  >
    <div
      className="movie-card__hero"
    >
      <div
        className="movie-card__bg"
      >
        <img
          alt="The Grand Budapest Hotel"
          src="img/bg-the-grand-budapest-hotel.jpg"
        />
      </div>
      <h1
        className="visually-hidden"
      >
        WTW
      </h1>
      <header
        className="page-header movie-card__head "
      >
        <div
          className="logo"
        >
          <a
            className="logo__link"
            href="/"
            onClick={[Function]}
          >
            <span
              className="logo__letter logo__letter--1"
            >
              W
            </span>
            <span
              className="logo__letter logo__letter--2"
            >
              T
            </span>
            <span
              className="logo__letter logo__letter--3"
            >
              W
            </span>
          </a>
        </div>
        <div
          className="user-block"
        >
          <a
            className="user-block__link"
            href="/login"
            onClick={[Function]}
          >
            Sign in
          </a>
        </div>
      </header>
      <div
        className="movie-card__wrap"
      >
        <div
          className="movie-card__desc"
        >
          <h2
            className="movie-card__title"
          >
            The Grand Budapest Hotel
          </h2>
          <p
            className="movie-card__meta"
          >
            <span
              className="movie-card__genre"
            >
              comedy
            </span>
            <span
              className="movie-card__year"
            >
              2010
            </span>
          </p>
          <div
            className="movie-card__buttons"
          >
            <button
              className="btn btn--play movie-card__button"
              onClick={[Function]}
              type="button"
            >
              <svg
                height="19"
                viewBox="0 0 19 19"
                width="19"
              >
                <use
                  href="#play-s"
                />
              </svg>
              <span>
                Play
              </span>
            </button>
            <button
              className="btn btn--list movie-card__button"
              onClick={[Function]}
              type="button"
            >
              <svg
                height="14"
                viewBox="0 0 18 14"
                width="18"
              >
                <use
                  xlinkHref="#in-list"
                />
              </svg>
              <span>
                My list
              </span>
            </button>
            <a
              className="btn movie-card__button"
              href="/films/1/review"
              onClick={[Function]}
            >
              Add review
            </a>
          </div>
        </div>
      </div>
    </div>
    <div
      className="movie-card__wrap movie-card__translate-top"
    >
      <div
        className="movie-card__info"
      >
        <div
          className="movie-card__poster movie-card__poster--big"
        >
          <img
            alt="The Grand Budapest Hotel"
            height="327"
            src="img/the-grand-budapest-hotel-poster.jpg"
            width="218"
          />
        </div>
        <div
          className="movie-card__desc"
        >
          <nav
            className="movie-nav movie-card__nav"
          >
            <ul
              className="movie-nav__list"
            >
              <li
                className="movie-nav__item movie-nav__item--active"
              >
                <a
                  className="movie-nav__link"
                  href="#"
                  onClick={[Function]}
                >
                  Overview
                </a>
              </li>
              <li
                className="movie-nav__item "
              >
                <a
                  className="movie-nav__link"
                  href="#"
                  onClick={[Function]}
                >
                  Detail
                </a>
              </li>
              <li
                className="movie-nav__item "
              >
                <a
                  className="movie-nav__link"
                  href="#"
                  onClick={[Function]}
                >
                  Review
                </a>
              </li>
            </ul>
          </nav>
          <div
            className="movie-rating"
          >
            <div
              className="movie-rating__score"
            >
              9.8
            </div>
            <p
              className="movie-rating__meta"
            >
              <span
                className="movie-rating__level"
              >
                Very good
              </span>
              <span
                className="movie-rating__count"
              >
                240
                 ratings
              </span>
            </p>
          </div>
          <div
            className="movie-card__text"
          >
            <p>
              In the 1930s, the Grand Budapest Hotel is a popular European ski resort, presided over by concierge Gustave H. (Ralph Fiennes). Zero, a junior lobby boy, becomes Gustaves friend and protege.  Gustave prides himself on providing first-class service to the hotels guests, including satisfying the sexual needs of the many elderly women who stay there. When one of Gustaves lovers dies mysteriously, Gustave finds himself the recipient of a priceless painting and the chief suspect in her murder.
            </p>
            <p
              className="movie-card__director"
            >
              <strong>
                Director: 
                Wes Andreson
              </strong>
            </p>
            <p
              className="movie-card__starring"
            >
              <strong>
                Starring: 
                Bill Murray, Edward Norton, Jude Law, Willem Dafoe, Saoirse Ronan, Tony Revoloru
                 and other
              </strong>
            </p>
          </div>
        </div>
      </div>
    </div>
  </section>,
  <div
    className="page-content"
  >
    <section
      className="catalog catalog--like-this"
    >
      <h2
        className="catalog__title"
      >
        More like this
      </h2>
      <div
        className="catalog__movies-list"
      >
        <article
          className="small-movie-card catalog__movies-card"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <div
            className="small-movie-card__image"
          >
            <video
              className="player__video"
              poster="img/bohemian-rhapsody.jpg"
            />
          </div>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
              href="/films/1"
              onClick={[Function]}
            >
              The Grand Budapest Hotel
            </a>
          </h3>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <div
            className="small-movie-card__image"
          >
            <video
              className="player__video"
              poster="img/bohemian-rhapsody.jpg"
            />
          </div>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
              href="/films/2"
              onClick={[Function]}
            >
              The Grand Budapest Hotel
            </a>
          </h3>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <div
            className="small-movie-card__image"
          >
            <video
              className="player__video"
              poster="img/bohemian-rhapsody.jpg"
            />
          </div>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
              href="/films/3"
              onClick={[Function]}
            >
              The Grand Budapest Hotel
            </a>
          </h3>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <div
            className="small-movie-card__image"
          >
            <video
              className="player__video"
              poster="img/bohemian-rhapsody.jpg"
            />
          </div>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
              href="/films/4"
              onClick={[Function]}
            >
              The Grand Budapest Hotel
            </a>
          </h3>
        </article>
      </div>
    </section>
    <footer
      className="page-footer"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link logo__link--light"
          href="/login"
          onClick={[Function]}
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="copyright"
      >
        <p>
          © 2019 What to watch Ltd.
        </p>
      </div>
    </footer>
  </div>,
]
`;
